---
title: "creating_biomaRt_query"
author: "Balint_Biro"
date: "1/20/2022"
output: github_document
---
```{r}
#There are two possible way to install biomaRt library
#install.packages('BiocManager',quietly=TRUE)
#BiocManager::install('biomaRt')
```

```{r}
#loading the library
library(biomaRt)
```

```{r}
#list all avaialable DBs
listEnsembl()
```
```{r}
#DB1-GENES
#select the desired dataset. To list the available datasets, first you have to
#specify your biomart (aka DB) and the preferred DB mirror (useast,uswest,asia
#and www)
ensembl <- useEnsembl(biomart='genes',
                      mirror='useast')
head(listDatasets(ensembl))
```
```{r}
#DB2-MOUSE STRAINS
#select the desired dataset. To list the available datasets, first you have to
#specify your biomart (aka DB) and the preferred DB mirror (useast,uswest,asia
#and www)
ensembl <- useEnsembl(biomart='mouse_strains',
                      mirror='useast')
head(listDatasets(ensembl))
```
```{r}
#DB3-SNPS
#select the desired dataset. To list the available datasets, first you have to
#specify your biomart (aka DB) and the preferred DB mirror (useast,uswest,asia
#and www)
ensembl <- useEnsembl(biomart='snps',
                      mirror='useast')
head(listDatasets(ensembl))
```
```{r}
#DB4-REGULATION
#select the desired dataset. To list the available datasets, first you have to
#specify your biomart (aka DB) and the preferred DB mirror (useast,uswest,asia
#and www)
ensembl <- useEnsembl(biomart='regulation',
                      mirror='useast')
head(listDatasets(ensembl))
```
```{r}
#After you select the dataset then you need to specify the Attributes. For a
#nice example I picked mouse strains and the genes of aj strain. Attributes are
#basically the features that you wanna have as return values (that you are
#interested in i.e. gene name, protein name etc.)
#all Attributes
ensembl <- useEnsembl(biomart='mouse_strains',
                      dataset='mmaj_gene_ensembl',
                      mirror='useast')
head(listAttributes(ensembl))
```
```{r}
#There is a bunch of different available attributes which can be quite
#tricky to go through if you are not sure what are you looking for so biomaRt
#offers a nice masking step to search for the appropriate attributes.
searchAttributes(mart=ensembl, pattern='go')
```
```{r}
#In the last step before the query you have to specify the Filters. The Filters
#are gonna determine the input for the query.
#all Filters
head(listAttributes(ensembl))
```
```{r}
#The situation is kinda same as it was at the Attributes when it comes to
#Filters so scrolling through a ton of Filters is filthy. But there is also a
#masking possibility
searchFilters(mart=ensembl, pattern='gene')
```
```{r}
#create the biomaRt query by using the previously detailed steps
#attributes are gonna be your return values (output)
#filters are the names of your input
#values are the actual inputs
#mart is the data where we specified the DB and Dataset
data <- getBM(attributes=c('external_gene_name','uniprotswissprot','go_id'),
                 filters=c('chromosomal_region'),
                 values=list('1:1230000:1450000'),
              mart=ensembl)
head(data)
```

